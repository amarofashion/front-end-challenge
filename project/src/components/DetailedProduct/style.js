import styled from 'styled-components';
import breakpoint from 'styled-components-breakpoint';

import colors from '../../theme/colors';

export const StyledWrapper = styled.div`
  display: flex;
  flex-flow: column;
  align-items: stretch;
  justify-content: stretch;
  height: auto;
  width: 100%;
  text-align: center;
  background: ${colors.light};

  .content {
    border: 1px solid ${colors.gray};
    position: relative;
    padding: 0 20px;
    height: auto;
    width: 100%;

    ${breakpoint('tablet')`
      height: 510px;
      width: 800px;
    `}

    &:hover {
      border-color: #ccc;
      transition: 0.2s all ease-in-out;
    }
  }

  .image {
    width: 100%;
    display: block;
    position: relative;

    ${breakpoint('tablet')`
      width: 50%;
      display: inline-block;
    `}

    img {
      width: 50%;
      object-fit: cover;
      margin: 10px auto;

      ${breakpoint('tablet')`
        width: 100%;
        object-fit: cover;
      `}
    }
  }

  .bottomContent {
    width: 100%;
    display: inline-block;
    padding: 5px 10px 15px;
    margin-top: 10px;

    ${breakpoint('tablet')`
      display: inline-block;
      vertical-align: top;
      width: 50%;
      margin-top: 50px;
    `}
  }

  .name {
    font-size: 20px;
    margin-bottom: 10px;
  }

  .prices {
    display: inline;
    margin-right: 10px;
  }

  .regularPrice {
    ${({ sale }) =>
      sale
        ? `
          font-size: 12px;
          position: relative;
          text-decoration: line-through;
          top: -2px;
        `
        : `
          font-size: 22px;
        `}
  }

  .actualPrice {
    color: ${colors.black};
    font-weight: normal;

    ${({ sale }) =>
      sale &&
      `
        color: ${colors.primary};
        font-weight: bold;
        font-size: 22px;
        position: relative;
        top: 3px;
      `}
  }

  .installments {
    color: #bbb;
    font-size: 12px;
    position: relative;
    top: -2px;
  }

  .submit {
    button {
      padding: 10px 20px;
      width: 100%;
      margin: 20px 0px;
      text-transform: uppercase;
      background: ${colors.primary};
      color: ${colors.light};
      font-weight: bold;
      border: 2px solid ${colors.primary};

      &:hover {
        cursor: pointer;
        border: 2px solid ${colors.dark};
        transition: 0.2s all ease-in-out;
      }
    }
  }
`;

export const StyledButton = styled.button`
  margin-top: 15px;
  text-align: center;
  width: 40px;
  display: inline-block;
  margin-right: 5px;
  padding: 5px 0px 10px;
  background: ${colors.light};
  font-size: 14px;
  font-weight: bold;

  border: ${({ selected }) =>
    selected ? `2px solid ${colors.dark}` : '2px solid #ccc'};

  &:hover {
    border-color: ${colors.dark};
    color: ${colors.dark};
    background: ${colors.light};
    cursor: pointer;
    transition: 0.2s all ease-in-out;
  }

  span {
    font-size: 10px;
  }
`;
