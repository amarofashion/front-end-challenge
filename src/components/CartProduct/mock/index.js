export default {
  name: 'VESTIDO TRANSPASSE BOW',
  style: '20002605',
  onSale: false,
  regularPrice: 'R$ 199,90',
  actualPrice: 'R$ 199,90',
  discountPercentage: '',
  installments: '3x R$ 66,63',
  image:
    'https://d3l7rqep7l31az.cloudfront.net/images/products/20002605_615_catalog_1.jpg?1460136912',
  sizes: [
    {
      available: false,
      size: 'PP',
      sku: '5807_343_0_PP',
    },
    {
      available: true,
      size: 'P',
      sku: '5807_343_0_P',
    },
    {
      available: true,
      size: 'M',
      sku: '5807_343_0_M',
    },
    {
      available: true,
      size: 'G',
      sku: '5807_343_0_G',
    },
    {
      available: false,
      size: 'GG',
      sku: '5807_343_0_GG',
    },
  ],
  selectedSize: {
    available: false,
    size: 'PP',
    sku: '5807_343_0_PP',
  },
  quantity: 1,
  quantityCallback: item => console.log(item),
  deleteCallback: item => console.log(item),
};
