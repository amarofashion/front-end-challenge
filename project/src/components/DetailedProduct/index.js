import React, { useState, useEffect } from 'react';
import { string, bool, shape, arrayOf, func } from 'prop-types';

import { StyledWrapper, StyledButton } from './style';

const propTypes = {
  image: string.isRequired,
  name: string.isRequired,
  regularPrice: string.isRequired,
  actualPrice: string.isRequired,
  installments: string.isRequired,
  onSale: bool.isRequired,
  sizes: arrayOf(
    shape({
      available: bool,
      size: string,
      sku: string,
    })
  ).isRequired,
  callback: func.isRequired,
};

function validate({ selectedSize }) {
  if (!selectedSize) return false;

  return true;
}

function DetailedProduct({
  image,
  name,
  regularPrice,
  onSale,
  actualPrice,
  installments,
  sizes,
  callback,
}) {
  const [selectedSize, setSelectedSize] = useState(null);
  const [error, setError] = useState(false);

  const addToCart = () => {
    if (validate({ selectedSize }) === false) return setError(true);

    setError(false);
    return callback(sizes[selectedSize - 1]);
  };

  useEffect(() => {
    setError(false);
  }, [selectedSize]);

  return (
    <StyledWrapper sale={onSale}>
      <div className='content'>
        <div className='image'>
          <img src={image} alt={`Foto do produto ${name}`} />
        </div>

        <div className='bottomContent'>
          <p className='name'>{name}</p>

          {onSale
            ? [
                <p key='regularPrice' className='regularPrice prices'>
                  {regularPrice}
                </p>,
                <p key='actualPrice' className='actualPrice prices'>
                  {actualPrice}
                </p>,
                <p key='installments' className='installments prices'>
                  {installments}
                </p>,
              ]
            : [
                <p key='regularPrice' className='regularPrice prices'>
                  {regularPrice}
                </p>,
                <p key='installments' className='installments prices'>
                  {installments}
                </p>,
              ]}

          <div>
            {sizes.map(
              ({ available, size }, index) =>
                available && (
                  <StyledButton
                    key={`size-${index}`}
                    onClick={() => setSelectedSize(index + 1)}
                    type='button'
                    selected={index + 1 === selectedSize}
                  >
                    <span>{size}</span>
                  </StyledButton>
                )
            )}
          </div>

          <div className='submit'>
            <button onClick={() => addToCart()} type='button'>
              adicionar ao carrinho
            </button>
          </div>

          <div>{error && <span>é necessário selecionar um tamanho</span>}</div>
        </div>
      </div>
    </StyledWrapper>
  );
}

DetailedProduct.propTypes = propTypes;

export default DetailedProduct;
