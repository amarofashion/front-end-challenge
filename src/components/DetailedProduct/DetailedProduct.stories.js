import React from 'react';

import DetailedProduct from '.';
import mock from './mock';

export default {
  component: DetailedProduct,
  title: 'DetailedProduct',
};

export const regular = () => (
  <div style={{ margin: 10 }}>
    <DetailedProduct {...mock} />
  </div>
);

export const onSale = () => (
  <div style={{ margin: 10 }}>
    <DetailedProduct {...mock} onSale />
  </div>
);
