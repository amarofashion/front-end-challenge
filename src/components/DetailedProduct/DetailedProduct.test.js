import React from 'react';
import { mount } from 'enzyme';

import DetailedProduct from '.';
import mock from './mock';

describe('<DetailedProduct />', () => {
  let wrapper;

  describe('with default props', () => {
    beforeAll(() => {
      wrapper = mount(<DetailedProduct {...mock} />);
    });

    it('shoud match snapshot with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
