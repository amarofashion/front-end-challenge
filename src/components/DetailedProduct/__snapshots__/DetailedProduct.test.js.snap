// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailedProduct /> with default props shoud match snapshot with default props 1`] = `
<DetailedProduct
  actualPrice="R$ 199,90"
  callback={[Function]}
  discountPercentage=""
  image="https://d3l7rqep7l31az.cloudfront.net/images/products/20002605_615_catalog_1.jpg?1460136912"
  installments="3x R$ 66,63"
  name="VESTIDO TRANSPASSE BOW"
  onSale={false}
  regularPrice="R$ 199,90"
  sizes={
    Array [
      Object {
        "available": false,
        "size": "PP",
        "sku": "5807_343_0_PP",
      },
      Object {
        "available": true,
        "size": "P",
        "sku": "5807_343_0_P",
      },
      Object {
        "available": true,
        "size": "M",
        "sku": "5807_343_0_M",
      },
      Object {
        "available": true,
        "size": "G",
        "sku": "5807_343_0_G",
      },
      Object {
        "available": false,
        "size": "GG",
        "sku": "5807_343_0_GG",
      },
    ]
  }
  style="20002605"
>
  <styled.div
    sale={false}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "kBWAJs",
            "rules": Array [
              "
  display: flex;
  flex-flow: column;
  align-items: stretch;
  justify-content: stretch;
  height: auto;
  width: 100%;
  text-align: center;
  background: ",
              "#fff",
              ";

  .content {
    border: 1px solid ",
              "#eee",
              ";
    position: relative;
    padding: 0 20px;
    height: auto;
    width: 100%;

    ",
              [Function],
              "

    &:hover {
      border-color: #ccc;
      transition: 0.2s all ease-in-out;
    }
  }

  .image {
    width: 100%;
    display: block;
    position: relative;

    ",
              [Function],
              "

    img {
      width: 50%;
      object-fit: cover;
      margin: 10px auto;

      ",
              [Function],
              "
    }
  }

  .bottomContent {
    width: 100%;
    display: inline-block;
    padding: 5px 10px 50px;
    margin-top: 10px;

    ",
              [Function],
              "
  }

  .name {
    font-size: 20px;
    margin-bottom: 10px;
  }

  .prices {
    display: inline;
    margin-right: 10px;
  }

  .regularPrice {
    ",
              [Function],
              "
  }

  .actualPrice {
    color: ",
              ";
    font-weight: normal;

    ",
              [Function],
              "
  }

  .installments {
    color: #bbb;
    font-size: 12px;
    position: relative;
    top: -2px;
  }

  .submit {
    button {
      padding: 10px 20px;
      width: 100%;
      margin: 20px 0px;
      text-transform: uppercase;
      background: ",
              "#E56A54",
              ";
      color: ",
              "#fff",
              ";
      font-weight: bold;
      border: 2px solid ",
              "#E56A54",
              ";

      &:hover {
        cursor: pointer;
        border: 2px solid ",
              "#000",
              ";
        transition: 0.2s all ease-in-out;
      }
    }
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      sale={false}
    >
      <div
        className="sc-bdVaJa kBWAJs"
      >
        <div
          className="content"
        >
          <div
            className="image"
          >
            <img
              alt="Foto do produto VESTIDO TRANSPASSE BOW"
              src="https://d3l7rqep7l31az.cloudfront.net/images/products/20002605_615_catalog_1.jpg?1460136912"
            />
          </div>
          <div
            className="bottomContent"
          >
            <p
              className="name"
            >
              VESTIDO TRANSPASSE BOW
            </p>
            <p
              className="regularPrice prices"
              key="regularPrice"
            >
              R$ 199,90
            </p>
            <p
              className="installments prices"
              key="installments"
            >
              3x R$ 66,63
            </p>
            <div>
              <styled.button
                data-cypress="size-item"
                key="size-1"
                onClick={[Function]}
                selected={false}
                type="button"
              >
                <StyledComponent
                  data-cypress="size-item"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bwzfXH",
                        "isStatic": false,
                        "lastClassName": "kKqhND",
                        "rules": Array [
                          "
  margin-top: 15px;
  text-align: center;
  width: 40px;
  display: inline-block;
  margin-right: 5px;
  padding: 5px 0px 10px;
  background: ",
                          "#fff",
                          ";
  font-size: 14px;
  font-weight: bold;

  border: ",
                          [Function],
                          ";

  &:hover {
    border-color: ",
                          "#000",
                          ";
    color: ",
                          "#000",
                          ";
    background: ",
                          "#fff",
                          ";
    cursor: pointer;
    transition: 0.2s all ease-in-out;
  }

  span {
    font-size: 10px;
  }
",
                        ],
                      },
                      "displayName": "styled.button",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bwzfXH",
                      "target": "button",
                      "toString": [Function],
                      "usesTheme": false,
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  onClick={[Function]}
                  selected={false}
                  type="button"
                >
                  <button
                    className="sc-bwzfXH kKqhND"
                    data-cypress="size-item"
                    onClick={[Function]}
                    selected={false}
                    type="button"
                  >
                    <span>
                      P
                    </span>
                  </button>
                </StyledComponent>
              </styled.button>
              <styled.button
                data-cypress="size-item"
                key="size-2"
                onClick={[Function]}
                selected={false}
                type="button"
              >
                <StyledComponent
                  data-cypress="size-item"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bwzfXH",
                        "isStatic": false,
                        "lastClassName": "kKqhND",
                        "rules": Array [
                          "
  margin-top: 15px;
  text-align: center;
  width: 40px;
  display: inline-block;
  margin-right: 5px;
  padding: 5px 0px 10px;
  background: ",
                          "#fff",
                          ";
  font-size: 14px;
  font-weight: bold;

  border: ",
                          [Function],
                          ";

  &:hover {
    border-color: ",
                          "#000",
                          ";
    color: ",
                          "#000",
                          ";
    background: ",
                          "#fff",
                          ";
    cursor: pointer;
    transition: 0.2s all ease-in-out;
  }

  span {
    font-size: 10px;
  }
",
                        ],
                      },
                      "displayName": "styled.button",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bwzfXH",
                      "target": "button",
                      "toString": [Function],
                      "usesTheme": false,
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  onClick={[Function]}
                  selected={false}
                  type="button"
                >
                  <button
                    className="sc-bwzfXH kKqhND"
                    data-cypress="size-item"
                    onClick={[Function]}
                    selected={false}
                    type="button"
                  >
                    <span>
                      M
                    </span>
                  </button>
                </StyledComponent>
              </styled.button>
              <styled.button
                data-cypress="size-item"
                key="size-3"
                onClick={[Function]}
                selected={false}
                type="button"
              >
                <StyledComponent
                  data-cypress="size-item"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bwzfXH",
                        "isStatic": false,
                        "lastClassName": "kKqhND",
                        "rules": Array [
                          "
  margin-top: 15px;
  text-align: center;
  width: 40px;
  display: inline-block;
  margin-right: 5px;
  padding: 5px 0px 10px;
  background: ",
                          "#fff",
                          ";
  font-size: 14px;
  font-weight: bold;

  border: ",
                          [Function],
                          ";

  &:hover {
    border-color: ",
                          "#000",
                          ";
    color: ",
                          "#000",
                          ";
    background: ",
                          "#fff",
                          ";
    cursor: pointer;
    transition: 0.2s all ease-in-out;
  }

  span {
    font-size: 10px;
  }
",
                        ],
                      },
                      "displayName": "styled.button",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bwzfXH",
                      "target": "button",
                      "toString": [Function],
                      "usesTheme": false,
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  onClick={[Function]}
                  selected={false}
                  type="button"
                >
                  <button
                    className="sc-bwzfXH kKqhND"
                    data-cypress="size-item"
                    onClick={[Function]}
                    selected={false}
                    type="button"
                  >
                    <span>
                      G
                    </span>
                  </button>
                </StyledComponent>
              </styled.button>
            </div>
            <div
              className="submit"
            >
              <button
                data-cypress="detailed-submit"
                onClick={[Function]}
                type="button"
              >
                adicionar ao carrinho
              </button>
            </div>
            <div />
          </div>
        </div>
      </div>
    </StyledComponent>
  </styled.div>
</DetailedProduct>
`;
